src/main/java/sandbox/JdbcSqlStorageIterator.java

```diff
@@ -1,28 +1,154 @@
+
 package sandbox;
 
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.commcare.modern.database.DatabaseHelper;
+import org.commcare.modern.database.TableBuilder;
+import org.javarosa.core.services.storage.IStorageIterator;
 import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.services.storage.StorageModifiedException;
+import org.javarosa.core.util.ArrayUtilities;
 
-import java.util.ArrayList;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.*;
 
 /**
- * Simple implemenation of IStorageIterator that is actually just backed by an ArrayList
- * since we don't want to deal with Connections or ResultSets lingering around.
+ * Iterator over JDBC ResultSet
  *
+ * @author ctsims
  * @author wspride
  */
-public class JdbcSqlStorageIterator <E extends Persistable> extends AbstractSqlIterator {
+public class JdbcSqlStorageIterator<T extends Persistable> implements IStorageIterator<T>, Iterator<T> {
+
+    protected PreparedStatement preparedStatement;
+    protected ResultSet resultSet;
+    private final Set<String> metaDataIndexSet;
+    SqliteIndexedStorageUtility<T> storage;
+    private HashMap<String, Integer> metaDataColumnMap = new HashMap<>();
 
-    public JdbcSqlStorageIterator(ArrayList<E> backingList) {
-        super(backingList);
+    private static final Log log = LogFactory.getLog(JdbcSqlStorageIterator.class);
+
+    public JdbcSqlStorageIterator(PreparedStatement preparedStatement,
+                                  ResultSet resultSet,
+                                  SqliteIndexedStorageUtility<T> storage,
+                                  String[] metaDataIndexSet) {
+        this.resultSet = resultSet;
+        this.storage = storage;
+        this.metaDataIndexSet = new HashSet<>(ArrayUtilities.toVector(metaDataIndexSet));
+        this.preparedStatement = preparedStatement;
+        try {
+            if(!resultSet.next()) {
+                resultSet.close();
+                preparedStatement.close();
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
     }
 
-    E getCurrentRecord() {
-        return ((E)backingList.get(index));
+    @Override
+    public int peekID() {
+        try {
+            return resultSet.getInt(DatabaseHelper.ID_COL);
+        } catch (SQLException e) {
```

Comment:
Wow, result sets... really suck. I can't believe this is a strongly typed exception that needs to be handled ~everywhere rather than an untyped exception.